src/poly/commitment/prover.rs

```diff
@@ -1,202 +1,183 @@
+    // probability.
+    let iota = *ChallengeScalar::<C, ()>::get(transcript);
+
+    // Challenge that ensures that the prover did not interfere with the U term
+    // in their commitments.
+    let z = *ChallengeScalar::<C, ()>::get(transcript);
+
+    // We'll be opening `s_poly_commitment * iota + P - [v] G_0` to ensure it
+    // has a root at zero.
+    let mut final_poly = s_poly * iota + px;
+    let v = eval_polynomial(&final_poly, x);
+    final_poly[0] = final_poly[0] - &v;
+    drop(px);
+    let blind = s_poly_blind * Blind(iota) + blind;
+    let mut blind = blind.0;
+    drop(s_poly_blind);
+
+    // Initialize the vector `a` as the coefficients of the polynomial,
+    // rounding up to the parameters.
+    let mut a = final_poly.values;
+    a.resize(params.n as usize, C::Scalar::zero());
+
+    // Initialize the vector `b` as the powers of `x`. The inner product of
+    // `a` and `b` is the evaluation of the polynomial at `x`.
+    let mut b = Vec::with_capacity(1 << params.k);
     {
-        let mut blind = blind.0;
-
-        // We're limited to polynomials of degree n - 1.
-        assert!(px.len() <= params.n as usize);
-
-        // Compute U
-        let u = {
-            let u_x = transcript.squeeze();
-            // y^2 = x^3 + B
-            let u_y2 = u_x.square() * &u_x + &C::b();
-            if let Some(u_y) = u_y2.deterministic_sqrt() {
-                C::from_xy(u_x, u_y).unwrap()
+        let mut cur = C::Scalar::one();
+        for _ in 0..(1 << params.k) {
+            b.push(cur);
+            cur *= &x;
+        }
+    }
+
+    // Initialize the vector `G` from the SRS. We'll be progressively
+    // collapsing this vector into smaller and smaller vectors until it is
+    // of length 1.
+    let mut g = params.g.clone();
+
+    // Perform the inner product argument, round by round.
+    for k in (1..=params.k).rev() {
+        let half = 1 << (k - 1); // half the length of `a`, `b`, `G`
+
+        // Compute L, R
+        //
+        // TODO: If we modify multiexp to take "extra" bases, we could speed
+        // this piece up a bit by combining the multiexps.
+        metrics::counter!("multiexp", 2, "val" => "l/r", "size" => format!("{}", half));
+        let l = best_multiexp(&a[0..half], &g[half..]);
+        let r = best_multiexp(&a[half..], &g[0..half]);
+        let value_l = compute_inner_product(&a[0..half], &b[half..]);
+        let value_r = compute_inner_product(&a[half..], &b[0..half]);
+        let mut l_randomness = C::Scalar::rand();
+        let r_randomness = C::Scalar::rand();
+        metrics::counter!("multiexp", 2, "val" => "l/r", "size" => "2");
+        let l = l + &best_multiexp(&[value_l * &z, l_randomness], &[params.u, params.h]);
+        let r = r + &best_multiexp(&[value_r * &z, r_randomness], &[params.u, params.h]);
+        let mut l = l.to_affine();
+        let r = r.to_affine();
+
+        let challenge = loop {
```

Comment:
We should be able to change the argument so that we aren't required to sample squares, to be more in line with the accumulation scheme paper. This would allow us to remove the entire `.fork()` and related APIs from the `transcript` module.

Reply:
I'll do this in a separate PR to make review simpler.